From 3cb18b4f91c10b769c11900dd6106deb0cbc88ac Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 3 Feb 2016 11:52:40 -0700
Subject: [PATCH] Remove duplicate hopper updates

Chests call update themselves, and mc  decided to throw in an ultimately useless duplicate of its own!
Bring it down to 1 per update by ignoring .update() during hopper manipulation

diff --git a/src/main/java/net/minecraft/server/IInventory.java b/src/main/java/net/minecraft/server/IInventory.java
index 52b8fc4..8398c34 100644
--- a/src/main/java/net/minecraft/server/IInventory.java
+++ b/src/main/java/net/minecraft/server/IInventory.java
@@ -12,6 +12,12 @@ public interface IInventory extends INamableTileEntity {
 
     ItemStack splitWithoutUpdate(int i);
 
+    // TacoSpigot start - add a 'setItem()' that doesn't update
+    default void setItem(int i, ItemStack stack, boolean shouldUpdate) {
+        setItem(i, stack);
+    }
+    // TacoSpigot end
+
     void setItem(int i, ItemStack itemstack);
 
     int getMaxStackSize();
diff --git a/src/main/java/net/minecraft/server/InventorySubcontainer.java b/src/main/java/net/minecraft/server/InventorySubcontainer.java
index 0e45148..5608009 100644
--- a/src/main/java/net/minecraft/server/InventorySubcontainer.java
+++ b/src/main/java/net/minecraft/server/InventorySubcontainer.java
@@ -143,13 +143,19 @@ public class InventorySubcontainer implements IInventory {
         }
     }
 
-    public void setItem(int i, ItemStack itemstack) {
+    // TacoSpigot start - add version that doesn't update
+    public void setItem(int i, ItemStack stack) {
+        setItem(i, stack, true);
+    }
+
+    public void setItem(int i, ItemStack itemstack, boolean shouldUpdate) {
+        // TacoSpigot end
         this.items[i] = itemstack;
         if (itemstack != null && itemstack.count > this.getMaxStackSize()) {
             itemstack.count = this.getMaxStackSize();
         }
 
-        this.update();
+        if (shouldUpdate) this.update(); // TacoSpigot
     }
 
     public int getSize() {
diff --git a/src/main/java/net/minecraft/server/TileEntityChest.java b/src/main/java/net/minecraft/server/TileEntityChest.java
index 5192d95..94354fb 100644
--- a/src/main/java/net/minecraft/server/TileEntityChest.java
+++ b/src/main/java/net/minecraft/server/TileEntityChest.java
@@ -92,13 +92,19 @@ public class TileEntityChest extends TileEntityContainer implements IInventory {
         }
     }
 
-    public void setItem(int i, ItemStack itemstack) {
+    // TacoSpigot start - add no update version of 'setItem()'
+    public void setItem(int i, ItemStack stack) {
+        setItem(i, stack, true);
+    }
+
+    public void setItem(int i, ItemStack itemstack, boolean shouldUpdate) {
+        // TacoSpigot end
         this.items[i] = itemstack;
         if (itemstack != null && itemstack.count > this.getMaxStackSize()) {
             itemstack.count = this.getMaxStackSize();
         }
 
-        this.update();
+        if (shouldUpdate) this.update(); // TacoSpigot
     }
 
     public String getName() {
diff --git a/src/main/java/net/minecraft/server/TileEntityDispenser.java b/src/main/java/net/minecraft/server/TileEntityDispenser.java
index 9be494b..1e6b628 100644
--- a/src/main/java/net/minecraft/server/TileEntityDispenser.java
+++ b/src/main/java/net/minecraft/server/TileEntityDispenser.java
@@ -98,13 +98,19 @@ public class TileEntityDispenser extends TileEntityContainer implements IInvento
         return i;
     }
 
-    public void setItem(int i, ItemStack itemstack) {
+    // TacoSpigot start - add a 'setItem()' that doesn't update
+    public void setItem(int i, ItemStack stack) {
+        setItem(i, stack, true);
+    }
+
+    public void setItem(int i, ItemStack itemstack, boolean shouldUpdate) {
+        // TacoSpigot end
         this.items[i] = itemstack;
         if (itemstack != null && itemstack.count > this.getMaxStackSize()) {
             itemstack.count = this.getMaxStackSize();
         }
 
-        this.update();
+        if (shouldUpdate) this.update(); // TacoSpigot
     }
 
     public int addItem(ItemStack itemstack) {
diff --git a/src/main/java/net/minecraft/server/TileEntityHopper.java b/src/main/java/net/minecraft/server/TileEntityHopper.java
index d0b448f..b0a3350 100644
--- a/src/main/java/net/minecraft/server/TileEntityHopper.java
+++ b/src/main/java/net/minecraft/server/TileEntityHopper.java
@@ -559,7 +559,7 @@ public class TileEntityHopper extends TileEntityContainer implements IHopper, IU
             boolean flag = false;
 
             if (itemstack1 == null) {
-                iinventory.setItem(i, itemstack);
+                iinventory.setItem(i, itemstack, false); // TacoSpigot - don't update inventory
                 itemstack = null;
                 flag = true;
             } else if (a(itemstack1, itemstack)) {
@@ -579,7 +579,7 @@ public class TileEntityHopper extends TileEntityContainer implements IHopper, IU
                         tileentityhopper.d(tileentityhopper.world.spigotConfig.hopperTransfer); // Spigot
                     }
 
-                    iinventory.update();
+                    // iinventory.update(); // TacoSpigot - remove duplicate
                 }
 
                 iinventory.update();
-- 
2.7.0

